crates/control/src/calibration_controller.rs

```diff
@@ -0,0 +1,270 @@
+                dispatch_time,
+                camera,
+            } => {
+                let calibration_response = match camera {
+                    CameraPosition::Top => {
+                        collect_filtered_values(&context.measurement_top, &dispatch_time)
+                    }
+                    CameraPosition::Bottom => {
+                        collect_filtered_values(&context.measurement_bottom, &dispatch_time)
+                    }
+                };
+
+                calibration_response.and_then(|response| {
+                    let goto_next_lookat = if let Some(measurement) = response.measurement {
+                        self.current_measurements.push(measurement);
+                        true
+                    } else {
+                        self.last_capture_retries += 1;
+                        self.last_capture_retries > *context.max_retries_per_capture
+                    };
+                    if goto_next_lookat {
+                        Some(self.try_transition_to_look_at(*context.cycle_time))
+                    } else {
+                        None
+                    }
+                })
+            }
+            CalibrationCommand::Process => {
+                // TODO Handle not enough measurements
+
+                let solved_result = solve(
+                    Corrections::default(),
+                    self.current_measurements.clone(),
+                    *context.field_dimensions,
+                );
+
+                self.last_calibration_corrections = Some(solved_result);
+                Some(CalibrationCommand::Finish)
+            }
+            CalibrationCommand::Finish => None,
+        }
+    }
+
+    fn try_transition_to_look_at(&mut self, dispatch_time: CycleTime) -> CalibrationCommand {
+        match self.get_next_look_at() {
+            Some((target, camera)) => CalibrationCommand::LookAt {
+                camera,
+                target,
+                dispatch_time,
+            },
+            None => CalibrationCommand::Process,
+        }
+    }
+
+    fn get_next_look_at(&mut self) -> Option<(Point2<Ground>, CameraPosition)> {
+        let current_index = self.current_look_at_index;
+        self.current_look_at_index += 1;
+        self.look_at_list.get(current_index).copied()
+    }
+}
+
+fn collect_filtered_values(
+    measurement_perception_input: &PerceptionInput<
+        Vec<Option<&CalibrationCaptureResponse<Measurement>>>,
+    >,
+    original_dispatch_time: &CycleTime,
+) -> Option<CalibrationCaptureResponse<Measurement>> {
+    measurement_perception_input
+        .persistent
+        .iter()
+        .flat_map(|(_cycle_timestamp, measurements)| measurements.iter())
+        .flatten()
+        .find(|&measurement| {
+            let CalibrationCaptureResponse {
+                dispatch_time,
+                measurement: _,
+            } = measurement;
+            original_dispatch_time.start_time == dispatch_time.start_time
```

Comment:
```suggestion
            original_dispatch_time.start_time == measurement.dispatch_time.start_time
```

Reply:
Uh that was a stupid one by me!  :3 fixed!